workflow-manager/bucket/bucket.go

```diff
@@ -123,7 +139,40 @@ func (b *Bucket) listFilesS3(ctx context.Context) ([]string, error) {
 	return output, nil
 }
 
-func (b *Bucket) listFilesGS(ctx context.Context) ([]string, error) {
+func (b *Bucket) writeTaskMarkerS3(ctx context.Context, marker string) error {
+	region, bucket, err := parseS3BucketName(b.bucketName)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("writing task marker to s3://%s/%s as %q", bucket, marker, b.identity)
+
+	if b.dryRun {
+		log.Printf("dry run, skipping marker write")
+		return nil
+	}
+
+	svc, err := b.s3Service(region)
+	if err != nil {
+		return err
+	}
+	input := &s3.PutObjectInput{
+		// Doesn't matter what the file contents are, but use the task name just
+		// in case S3 balks at an empty body
+		Body:   aws.ReadSeekCloser(strings.NewReader(marker)),
+		Bucket: aws.String(bucket),
+		Key:    aws.String(marker),
+	}
+
+	// Deliberately ignore the result, we only care if the write succeeds
+	if _, err := svc.PutObject(input); err != nil {
+		return fmt.Errorf("storage.PutObject: %w", err)
+	}
+
+	return nil
+}
+
+func (b *Bucket) gcsClient(ctx context.Context) (*storage.Client, error) {
```

Comment:
I'd drop passing ctx here and use 
```
ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
defer cancel()
```
to get the context for obtaining the client. The NewClient context is just needed for the client creation, for subsequent calls you use a new / different context.